File: function_app.py
import azure.functions as func
import datetime
import requests
import json
from azure.identity import DefaultAzureCredential

app = func.FunctionApp(http_auth_level=func.AuthLevel.ANONYMOUS)


@app.route(route="costs")
def get_costs(req: func.HttpRequest) -> func.HttpResponse:
    try:
        credential = DefaultAzureCredential()
        token = credential.get_token("https://management.azure.com/.default").token

        subscription_id = "71e85381-5707-48bd-9e75-199a87e9705e"

        today = datetime.date.today()
        start = today.replace(day=1).isoformat()

        url = (
            f"https://management.azure.com/subscriptions/{subscription_id}"
            f"/providers/Microsoft.CostManagement/query?api-version=2023-03-01"
        )

        headers = {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json"
        }

        body = {
            "type": "ActualCost",
            "timeframe": "Custom",
            "timePeriod": {
                "from": start,
                "to": today.isoformat()
            },
            "dataset": {
                "granularity": "Daily",
                "aggregation": {
                    "totalCost": {
                        "name": "PreTaxCost",
                        "function": "Sum"
                    }
                }
            }
        }

        response = requests.post(url, headers=headers, json=body)
        data = response.json()

        rows = data.get("properties", {}).get("rows", [])

        daily = []
        total = 0.0

        for r in rows:
            cost = float(r[1])
            total += cost
            daily.append({
                "date": str(r[0]),
                "cost": round(cost, 2)
            })

        # Simple projection
        days = len(daily)
        avg = total / days if days else 0

        days_in_month = (datetime.date(today.year, today.month % 12 + 1, 1)
                         - datetime.timedelta(days=1)).day

        projected = avg * days_in_month

        result = {
            "month_to_date_total": round(total, 2),
            "average_daily_cost": round(avg, 2),
            "projected_month_total": round(projected, 2),
            "daily_breakdown": daily
        }

        return func.HttpResponse(
            json.dumps(result),
            mimetype="application/json",
            status_code=200
        )

    except Exception as e:
        return func.HttpResponse(
            json.dumps({"error": str(e)}),
            status_code=500
        )

File: function_app.js
const { app } = require('@azure/functions');
const { DefaultAzureCredential } = require('@azure/identity');
const axios = require('axios');

app.http('costs', {
    methods: ['GET'],
    authLevel: 'anonymous',

    handler: async (request, context) => {

        try {

            const subscriptionId = '71e85381-5707-48bd-9e75-199a87e9705e';

            const credential = new DefaultAzureCredential();

            const token = await credential.getToken(
                'https://management.azure.com/.default'
            );

            const today = new Date();

            const startDate =
                `${today.getUTCFullYear()}-${String(today.getUTCMonth() + 1).padStart(2, '0')}-01`;

            const endDate =
                `${today.getUTCFullYear()}-${String(today.getUTCMonth() + 1).padStart(2, '0')}-${String(today.getUTCDate()).padStart(2, '0')}`;

            const url =
                `https://management.azure.com/subscriptions/${subscriptionId}` +
                `/providers/Microsoft.CostManagement/query` +
                `?api-version=2023-03-01`;

            const body = {
                type: 'ActualCost',
                timeframe: 'Custom',
                timePeriod: {
                    from: startDate,
                    to: endDate
                },
                dataset: {
                    granularity: 'Daily',
                    aggregation: {
                        totalCost: {
                            name: 'PreTaxCost',
                            function: 'Sum'
                        }
                    }
                }
            };

            const response = await axios.post(
                url,
                body,
                {
                    headers: {
                        Authorization: `Bearer ${token.token}`,
                        'Content-Type': 'application/json'
                    }
                }
            );

            const rows =
                response.data?.properties?.rows || [];

            let totalCost = 0;

            rows.forEach(row => {
                totalCost += Number(row[0]);
            });

            const dayOfMonth = today.getUTCDate();

            const projectedCost =
                dayOfMonth > 0
                    ? (totalCost / dayOfMonth) * 30
                    : totalCost;

            return {
                jsonBody: {
                    subscriptionId,
                    monthToDateSpend: `£${totalCost.toFixed(2)}`,
                    projectedMonthEndSpend: `£${projectedCost.toFixed(2)}`,
                    currency: 'GBP',
                    daysElapsed: dayOfMonth
                }
            };

        } catch (err) {

            return {
                status: 500,
                jsonBody: {
                    error: err.message,
                    details: err.response?.data || null
                }
            };

        }
    }
});